# Swing Vision Transformation
#### Converting SwingVision data into UCLA Tennis Consulting format
#### Run all cells ONCE; restart Kernel and Run All again if needed

#### TODO
- add firstServeLocation and isLet Columns
- Classification Models for isDropshot, isLob, isApproach - Leo's team

## Notebook Start

In [7]:
import pandas as pd
import numpy as np
import os 
import re

# Option to display max rows/columns
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

### Load in data

In [8]:
# Input file name here
your_file_name = 'GianlucaBallotta_MaxWestphal_Columbia.xlsx'
swing_data = pd.read_excel(your_file_name, sheet_name='Shots')
swing_data.shape

(545, 25)

In [9]:
swing_data.head()

Player  Shot          Type    Stroke     Spin  Speed (MPH)  \
0  Gianluca Ballotta     1   first_serve     Serve    Slice    98.378799   
1       Max Westphal     2  first_return  Backhand  Topspin    33.853065   
2       Max Westphal     0          none      Feed     Flat    18.186579   
3  Gianluca Ballotta     1   first_serve     Serve    Slice    93.113174   
4       Max Westphal     2  first_return  Backhand  Topspin    41.675724   

   Point  Game  Set Bounce Depth  Bounce Zone Bounce Side  Bounce (x)  \
0      1     1    1        short        deuce         far   -3.095711   
1      1     1    1        short  deuce_alley         far   -4.698338   
2      2     1    1        short        deuce        near    0.734488   
3      2     1    1        short           ad         far    1.086910   
4      2     1    1        short     ad_alley        near   -4.890435   

   Bounce (y) Hit Depth     Hit Zone Hit Side   Hit (x)    Hit (y)   Hit (z)  \
0   18.234331      deep        deuce     near  0.698740  -0.083398  2.605832   
1   17.903118       out  deuce_alley      far -5.395302  25.629480  0.802931   
2    6.336324     short           ad      far  0.806848  14.838730  2.732491   
3   16.193851      deep           ad     near -1.533032   0.108099  2.562802   
4    7.557028       out           ad      far  2.795515  26.246744  1.041161   

     Direction Result  Favorited Start Time  Video Time  
0     out wide     In      False   03:56:43        4.22  
1  cross court    Net      False   03:56:44        4.99  
2          ---     In      False   03:56:52       12.67  
3   down the T     In      False   03:57:06       26.92  
4    inside in    Out      False   03:57:07       27.75

### Clean swing vision data

In [11]:
swing_data.query('Stroke == "Feed"').shape

(48, 25)

In [12]:
swing_data.query('Shot == 0').shape

(40, 25)

In [13]:
swing_data.query('Type == "none"').shape

(68, 25)

#### Remove all the rows where it is a feed

In [14]:
swing_data.query('Type == "none"')

Player  Shot  Type    Stroke     Spin  Speed (MPH)  Point  \
2         Max Westphal     0  none      Feed     Flat    18.186579      2   
30   Gianluca Ballotta     0  none      Feed     Flat    18.186579      8   
31        Max Westphal     1  none      Feed     Flat    21.650688      8   
33   Gianluca Ballotta     3  none  Backhand     Flat    34.404015      8   
39   Gianluca Ballotta     0  none      Feed     Flat    22.516716     10   
41        Max Westphal     0  none      Feed     Flat    25.114799     11   
42        Max Westphal     1  none  Backhand  Topspin    46.298294     11   
48        Max Westphal     1  none  Forehand    Slice    35.318470     13   
49   Gianluca Ballotta     2  none  Backhand  Topspin    75.164177     13   
50        Max Westphal     3  none  Forehand  Topspin    26.694437     13   
68        Max Westphal     1  none  Forehand     Flat    42.751759     15   
72        Max Westphal     0  none      Feed     Flat    19.918634     16   
73        Max Westphal     1  none  Backhand  Topspin    40.318375     16   
85   Gianluca Ballotta     0  none      Feed     Flat    20.784662     20   
89   Gianluca Ballotta     0  none      Feed     Flat    24.248772     21   
95   Gianluca Ballotta     2  none  Backhand  Topspin    42.285610     22   
96   Gianluca Ballotta     3  none      Feed     Flat    22.516716     22   
102       Max Westphal     0  none      Feed     Flat    22.516716     23   
118       Max Westphal     0  none      Feed  Topspin    20.784662     24   
126       Max Westphal     1  none  Forehand     Flat    41.820976     25   
146       Max Westphal     0  none      Feed     Flat    21.650688     29   
147  Gianluca Ballotta     1  none  Backhand  Topspin    50.158554     29   
149       Max Westphal     1  none  Backhand     Flat    41.011642     30   
150  Gianluca Ballotta     2  none      Feed     Flat    12.709421     30   
153  Gianluca Ballotta     0  none      Feed     Flat    19.052607     31   
154       Max Westphal     1  none  Forehand  Topspin    45.594936     31   
184       Max Westphal     0  none      Feed     Flat    19.918634     37   
185       Max Westphal     1  none      Feed     Flat    17.320551     37   
230       Max Westphal     0  none      Feed     Flat    19.052607     42   
235       Max Westphal     0  none      Feed     Flat    11.559569     43   
241  Gianluca Ballotta     0  none      Feed     Flat    19.966738     44   
249  Gianluca Ballotta     0  none      Feed     Flat    17.320551     46   
254       Max Westphal     0  none      Feed     Flat    24.248772     47   
260       Max Westphal     0  none      Feed     Flat    24.248772     48   
266  Gianluca Ballotta     0  none      Feed     Flat    20.241814     49   
302  Gianluca Ballotta     0  none      Feed     Flat    22.516716     55   
308       Max Westphal     0  none      Feed     Flat    15.731391     57   
314       Max Westphal     0  none      Feed     Flat    20.784662     58   
316       Max Westphal     2  none  Forehand     Flat    32.941231     58   
320       Max Westphal     0  none      Feed     Flat    19.918634     59   
333       Max Westphal     0  none      Feed     Flat    23.653185     62   
343  Gianluca Ballotta     0  none      Feed  Topspin    17.320551     64   
345  Gianluca Ballotta     0  none      Feed     Flat    17.320551     65   
354       Max Westphal     0  none      Feed     Flat    23.382744     67   
355       Max Westphal     1  none      Feed     Flat    25.114799     67   
357       Max Westphal     3  none  Forehand    Slice    39.742020     67   
371       Max Westphal     0  none      Feed     Flat    23.382744     71   
372       Max Westphal     1  none      Feed     Flat    20.784662     71   
389       Max Westphal     0  none      Feed     Flat    20.784662     74   
393  Gianluca Ballotta     1  none  Backhand    Slice    61.518326     75   
406  Gianluca Ballotta     0  none      Feed     Flat    24.580153     76   
420       Ma

In [15]:
rows_to_drop = swing_data.query('Type == "none"').index
swing_data = swing_data.drop(rows_to_drop)
swing_data = swing_data.reset_index(drop=True) # Important to reindex to avoid missing indicies
swing_data.shape

(477, 25)

In [16]:
swing_data.head()

Player  Shot            Type    Stroke     Spin  Speed (MPH)  \
0  Gianluca Ballotta     1     first_serve     Serve    Slice    98.378799   
1       Max Westphal     2    first_return  Backhand  Topspin    33.853065   
2  Gianluca Ballotta     1     first_serve     Serve    Slice    93.113174   
3       Max Westphal     2    first_return  Backhand  Topspin    41.675724   
4  Gianluca Ballotta     3  serve_plus_one  Backhand  Topspin    33.080677   

   Point  Game  Set Bounce Depth  Bounce Zone Bounce Side  Bounce (x)  \
0      1     1    1        short        deuce         far   -3.095711   
1      1     1    1        short  deuce_alley         far   -4.698338   
2      2     1    1        short           ad         far    1.086910   
3      2     1    1        short     ad_alley        near   -4.890435   
4      2     1    1         deep        deuce         far   -1.746811   

   Bounce (y) Hit Depth     Hit Zone Hit Side   Hit (x)    Hit (y)   Hit (z)  \
0   18.234331      deep        deuce     near  0.698740  -0.083398  2.605832   
1   17.903118       out  deuce_alley      far -5.395302  25.629480  0.802931   
2   16.193851      deep           ad     near -1.533032   0.108099  2.562802   
3    7.557028       out           ad      far  2.795515  26.246744  1.041161   
4   19.188076     short           ad     near -1.135870  10.734202  3.925683   

       Direction Result  Favorited Start Time  Video Time  
0       out wide     In      False   03:56:43    4.220000  
1    cross court    Net      False   03:56:44    4.990000  
2     down the T     In      False   03:57:06   26.920000  
3      inside in    Out      False   03:57:07   27.750000  
4  down the line     In      False   03:57:10   30.870001

### Load in Points data

In [17]:
swing_data_points = pd.read_excel(your_file_name, sheet_name='Points')
swing_data_points.shape

(101, 15)

In [18]:
def create_point(server, player1score, player2score):
    if server == "host":
        return str(player1score) + "-" + str(player2score)
    else:
        return str(player2score) + "-" + str(player1score)
    
swing_data_points['pointScore'] = swing_data_points.apply(lambda x: create_point(x['Match Server'], x['Host Game Score'], x['Guest Game Score']), axis=1)

In [19]:
swing_data_points = swing_data_points.rename(columns={'Break Point' : 'isBreakPoint'})
swing_data_points['isBreakPoint'] = swing_data_points['isBreakPoint'].replace(False, '')
swing_data_points['isBreakPoint'] = swing_data_points['isBreakPoint'].replace(True, 1)

In [20]:
swing_data_points = swing_data_points[['Point', 'pointScore', 'isBreakPoint']]
swing_data_points.head()

Point pointScore isBreakPoint
0      1        0-0             
1      2       15-0             
2      3       30-0             
3      4       40-0             
4      5      40-15

In [21]:
swing_data = pd.merge(swing_data, swing_data_points, on='Point')

### Load in Games data

In [22]:
swing_data_games = pd.read_excel(your_file_name, sheet_name='Games')
swing_data_games.shape

(18, 9)

In [23]:
swing_data_games.head()

Game  Set Server  Host Set Score  Guest Set Score Game Winner Start Time  \
0     1    1   host               0                0        host   03:56:45   
1     2    1  guest               1                0       guest   03:58:38   
2     3    1   host               1                1       guest   04:02:54   
3     4    1  guest               1                2        host   04:07:48   
4     5    1   host               2                2        host   04:10:26   

   Video Time    Duration  
0    6.220000  113.129997  
1  119.349998  255.160004  
2  374.519989  294.600006  
3  669.109985  157.600006  
4  826.710022  233.949997

In [24]:
def create_game(player1game, player2game):
        return str(player1game) + "-" + str(player2game)

    
swing_data_games['gameScore'] = swing_data_games.apply(lambda x: create_game(x['Host Set Score'], x['Guest Set Score']),  axis=1)

In [25]:
swing_data_games

Game  Set Server  Host Set Score  Guest Set Score Game Winner Start Time  \
0      1    1   host               0                0        host   03:56:45   
1      2    1  guest               1                0       guest   03:58:38   
2      3    1   host               1                1       guest   04:02:54   
3      4    1  guest               1                2        host   04:07:48   
4      5    1   host               2                2        host   04:10:26   
5      6    1  guest               3                2       guest   04:14:20   
6      7    1   host               3                3        host   04:19:44   
7      8    1  guest               4                3       guest   04:24:36   
8      9    1   host               4                4       guest   04:27:53   
9     10    1  guest               4                5       guest   04:31:48   
10    11    2   host               0                0       guest   04:35:21   
11    12    2  guest               0                1       guest   04:41:11   
12    13    2   host               0                2        host   04:43:49   
13    14    2  guest               1                2       guest   04:48:23   
14    15    2   host               1                3       guest   04:52:18   
15    16    2  guest               1                4       guest   04:56:41   
16    17    2   host               1                5        host   05:00:54   
17    18    2  guest               2                5       guest   05:04:53   

     Video Time    Duration gameScore  
0      6.220000  113.129997       0-0  
1    119.349998  255.160004       1-0  
2    374.519989  294.600006       1-1  
3    669.109985  157.600006       1-2  
4    826.710022  233.949997       2-2  
5   1060.660034  324.579987       3-2  
6   1385.239990  291.410004       3-3  
7   1676.650024  197.199997       4-3  
8   1873.849976  234.800003       4-4  
9   2108.639893  212.800003       4-5  
10  2321.439941  350.700012       0-0  
11  2672.139893  157.500000       0-1  
12  2829.629883  274.359985       0-2  
13  3104.000000  235.330002       1-2  
14  3339.330078  262.100006       1-3  
15  3601.419922  253.600006       1-4  
16  3855.020020  239.199997       1-5  
17  4094.219971  176.869995       2-5

In [26]:
swing_data_games = swing_data_games[['Game', 'gameScore']]

In [27]:
swing_data = pd.merge(swing_data, swing_data_games, on="Game")

In [28]:
swing_data.head()

Player  Shot            Type    Stroke     Spin  Speed (MPH)  \
0  Gianluca Ballotta     1     first_serve     Serve    Slice    98.378799   
1       Max Westphal     2    first_return  Backhand  Topspin    33.853065   
2  Gianluca Ballotta     1     first_serve     Serve    Slice    93.113174   
3       Max Westphal     2    first_return  Backhand  Topspin    41.675724   
4  Gianluca Ballotta     3  serve_plus_one  Backhand  Topspin    33.080677   

   Point  Game  Set Bounce Depth  Bounce Zone Bounce Side  Bounce (x)  \
0      1     1    1        short        deuce         far   -3.095711   
1      1     1    1        short  deuce_alley         far   -4.698338   
2      2     1    1        short           ad         far    1.086910   
3      2     1    1        short     ad_alley        near   -4.890435   
4      2     1    1         deep        deuce         far   -1.746811   

   Bounce (y) Hit Depth     Hit Zone Hit Side   Hit (x)    Hit (y)   Hit (z)  \
0   18.234331      deep        deuce     near  0.698740  -0.083398  2.605832   
1   17.903118       out  deuce_alley      far -5.395302  25.629480  0.802931   
2   16.193851      deep           ad     near -1.533032   0.108099  2.562802   
3    7.557028       out           ad      far  2.795515  26.246744  1.041161   
4   19.188076     short           ad     near -1.135870  10.734202  3.925683   

       Direction Result  Favorited Start Time  Video Time pointScore  \
0       out wide     In      False   03:56:43    4.220000        0-0   
1    cross court    Net      False   03:56:44    4.990000        0-0   
2     down the T     In      False   03:57:06   26.920000       15-0   
3      inside in    Out      False   03:57:07   27.750000       15-0   
4  down the line     In      False   03:57:10   30.870001       15-0   

  isBreakPoint gameScore  
0                    0-0  
1                    0-0  
2                    0-0  
3                    0-0  
4                    0-0

### Load in Sets data

In [29]:
swing_data_sets = pd.read_excel(your_file_name, sheet_name='Sets')
swing_data_sets.shape

(2, 10)

In [30]:
swing_data_sets

Set  Host Score  Guest Score  Host Tiebreak Score  Guest Tiebreak Score  \
0    1           4            6                    0                     0   
1    2           2            6                    0                     0   

  Set Winner  Super Tiebreak Start Time   Video Time     Duration  
0      guest           False   03:56:45     6.220000  2315.219971  
1      guest           False   04:35:21  2321.439941  1949.640015

In [31]:
host_set_score = 0
guest_set_score = 0

def create_set(set_winner):
        global host_set_score, guest_set_score  # Declare global variables
        if set_winner == "host":
                host_set_score += 1
        else:        
                guest_set_score += 1
        
        return str(host_set_score) + "-" + str(guest_set_score)

swing_data_sets['setScore'] = None
swing_data_sets.at[0, 'setScore'] = "0-0"
    
swing_data_sets.iloc[1:, swing_data_sets.columns.get_loc('setScore')] = swing_data_sets.iloc[1:].apply(lambda x: create_set(x['Set Winner']),  axis=1)


In [32]:
swing_data_sets = swing_data_sets[['Set', 'setScore']]


In [33]:
swing_data = pd.merge(swing_data, swing_data_sets, on="Set")
swing_data.head()

Player  Shot            Type    Stroke     Spin  Speed (MPH)  \
0  Gianluca Ballotta     1     first_serve     Serve    Slice    98.378799   
1       Max Westphal     2    first_return  Backhand  Topspin    33.853065   
2  Gianluca Ballotta     1     first_serve     Serve    Slice    93.113174   
3       Max Westphal     2    first_return  Backhand  Topspin    41.675724   
4  Gianluca Ballotta     3  serve_plus_one  Backhand  Topspin    33.080677   

   Point  Game  Set Bounce Depth  Bounce Zone Bounce Side  Bounce (x)  \
0      1     1    1        short        deuce         far   -3.095711   
1      1     1    1        short  deuce_alley         far   -4.698338   
2      2     1    1        short           ad         far    1.086910   
3      2     1    1        short     ad_alley        near   -4.890435   
4      2     1    1         deep        deuce         far   -1.746811   

   Bounce (y) Hit Depth     Hit Zone Hit Side   Hit (x)    Hit (y)   Hit (z)  \
0   18.234331      deep        deuce     near  0.698740  -0.083398  2.605832   
1   17.903118       out  deuce_alley      far -5.395302  25.629480  0.802931   
2   16.193851      deep           ad     near -1.533032   0.108099  2.562802   
3    7.557028       out           ad      far  2.795515  26.246744  1.041161   
4   19.188076     short           ad     near -1.135870  10.734202  3.925683   

       Direction Result  Favorited Start Time  Video Time pointScore  \
0       out wide     In      False   03:56:43    4.220000        0-0   
1    cross court    Net      False   03:56:44    4.990000        0-0   
2     down the T     In      False   03:57:06   26.920000       15-0   
3      inside in    Out      False   03:57:07   27.750000       15-0   
4  down the line     In      False   03:57:10   30.870001       15-0   

  isBreakPoint gameScore setScore  
0                    0-0      0-0  
1                    0-0      0-0  
2                    0-0      0-0  
3                    0-0      0-0  
4                    0-0      0-0

### Create shot data csv

In [34]:
# Check existing columns
swing_data.columns

Index(['Player', 'Shot', 'Type', 'Stroke', 'Spin', 'Speed (MPH)', 'Point',
       'Game', 'Set', 'Bounce Depth', 'Bounce Zone', 'Bounce Side',
       'Bounce (x)', 'Bounce (y)', 'Hit Depth', 'Hit Zone', 'Hit Side',
       'Hit (x)', 'Hit (y)', 'Hit (z)', 'Direction', 'Result', 'Favorited',
       'Start Time', 'Video Time', 'pointScore', 'isBreakPoint', 'gameScore',
       'setScore'],
      dtype='object')

In [35]:
swing_data.head()

Player  Shot            Type    Stroke     Spin  Speed (MPH)  \
0  Gianluca Ballotta     1     first_serve     Serve    Slice    98.378799   
1       Max Westphal     2    first_return  Backhand  Topspin    33.853065   
2  Gianluca Ballotta     1     first_serve     Serve    Slice    93.113174   
3       Max Westphal     2    first_return  Backhand  Topspin    41.675724   
4  Gianluca Ballotta     3  serve_plus_one  Backhand  Topspin    33.080677   

   Point  Game  Set Bounce Depth  Bounce Zone Bounce Side  Bounce (x)  \
0      1     1    1        short        deuce         far   -3.095711   
1      1     1    1        short  deuce_alley         far   -4.698338   
2      2     1    1        short           ad         far    1.086910   
3      2     1    1        short     ad_alley        near   -4.890435   
4      2     1    1         deep        deuce         far   -1.746811   

   Bounce (y) Hit Depth     Hit Zone Hit Side   Hit (x)    Hit (y)   Hit (z)  \
0   18.234331      deep        deuce     near  0.698740  -0.083398  2.605832   
1   17.903118       out  deuce_alley      far -5.395302  25.629480  0.802931   
2   16.193851      deep           ad     near -1.533032   0.108099  2.562802   
3    7.557028       out           ad      far  2.795515  26.246744  1.041161   
4   19.188076     short           ad     near -1.135870  10.734202  3.925683   

       Direction Result  Favorited Start Time  Video Time pointScore  \
0       out wide     In      False   03:56:43    4.220000        0-0   
1    cross court    Net      False   03:56:44    4.990000        0-0   
2     down the T     In      False   03:57:06   26.920000       15-0   
3      inside in    Out      False   03:57:07   27.750000       15-0   
4  down the line     In      False   03:57:10   30.870001       15-0   

  isBreakPoint gameScore setScore  
0                    0-0      0-0  
1                    0-0      0-0  
2                    0-0      0-0  
3                    0-0      0-0  
4                    0-0      0-0

In [36]:
# add in all desired column labels, with swingvision labels at end

columm_names = (['pointScore', 'gameScore', 'setScore',
                'isPointStart', 'pointStartTime', 'isPointEnd', 'pointEndTime','pointNumber',
                'isBreakPoint','shotInRally','side','serverName',
                'serverFarNear','firstServeIn','firstServeZone',
                'firstServeXCoord','firstServeYCoord',
                'secondServeIn','secondServeZone','secondServeXCoord',
                'secondServeYCoord','isAce','shotContactX',
                'shotContactY','shotDirection','shotFhBh',
                'isSlice','isVolley','isOverhead','isApproach','isDropshot', 'isLet',
                'isExcitingPoint','atNetPlayer1','atNetPlayer2','isLob',
                'shotLocationX','shotLocationY','isWinner','isErrorWideR', 'isErrorWideL',
                'isErrorNet','isErrorLong','clientTeam',
                'Date', 'Division', 'Event', 'lineupPosition','matchDetails',
                'matchVenue' , 'opponentTeam', 
                'player1Name', 'player2Name','player1Hand','player2Hand',
            'Round','Surface','Notes'])

shot_data = pd.DataFrame(columns=columm_names)
shot_data

Empty DataFrame
Columns: [pointScore, gameScore, setScore, isPointStart, pointStartTime, isPointEnd, pointEndTime, pointNumber, isBreakPoint, shotInRally, side, serverName, serverFarNear, firstServeIn, firstServeZone, firstServeXCoord, firstServeYCoord, secondServeIn, secondServeZone, secondServeXCoord, secondServeYCoord, isAce, shotContactX, shotContactY, shotDirection, shotFhBh, isSlice, isVolley, isOverhead, isApproach, isDropshot, isLet, isExcitingPoint, atNetPlayer1, atNetPlayer2, isLob, shotLocationX, shotLocationY, isWinner, isErrorWideR, isErrorWideL, isErrorNet, isErrorLong, clientTeam, Date, Division, Event, lineupPosition, matchDetails, matchVenue, opponentTeam, player1Name, player2Name, player1Hand, player2Hand, Round, Surface, Notes]
Index: []

### Score Columns

In [37]:
shot_data['pointScore'] = swing_data['pointScore']
shot_data['gameScore'] = swing_data['gameScore']
shot_data['setScore'] = swing_data['setScore']

In [38]:
shot_data.head(10)

pointScore gameScore setScore isPointStart pointStartTime isPointEnd  \
0        0-0       0-0      0-0          NaN            NaN        NaN   
1        0-0       0-0      0-0          NaN            NaN        NaN   
2       15-0       0-0      0-0          NaN            NaN        NaN   
3       15-0       0-0      0-0          NaN            NaN        NaN   
4       15-0       0-0      0-0          NaN            NaN        NaN   
5       30-0       0-0      0-0          NaN            NaN        NaN   
6       30-0       0-0      0-0          NaN            NaN        NaN   
7       30-0       0-0      0-0          NaN            NaN        NaN   
8       30-0       0-0      0-0          NaN            NaN        NaN   
9       40-0       0-0      0-0          NaN            NaN        NaN   

  pointEndTime pointNumber isBreakPoint shotInRally side serverName  \
0          NaN         NaN          NaN         NaN  NaN        NaN   
1          NaN         NaN          NaN         NaN  NaN        NaN   
2          NaN         NaN          NaN         NaN  NaN        NaN   
3          NaN         NaN          NaN         NaN  NaN        NaN   
4          NaN         NaN          NaN         NaN  NaN        NaN   
5          NaN         NaN          NaN         NaN  NaN        NaN   
6          NaN         NaN          NaN         NaN  NaN        NaN   
7          NaN         NaN          NaN         NaN  NaN        NaN   
8          NaN         NaN          NaN         NaN  NaN        NaN   
9          NaN         NaN          NaN         NaN  NaN        NaN   

  serverFarNear firstServeIn firstServeZone firstServeXCoord firstServeYCoord  \
0           NaN          NaN            NaN              NaN              NaN   
1           NaN          NaN            NaN              NaN              NaN   
2           NaN          NaN            NaN              NaN              NaN   
3           NaN          NaN            NaN              NaN              NaN   
4           NaN          NaN            NaN              NaN              NaN   
5           NaN          NaN            NaN              NaN              NaN   
6           NaN          NaN            NaN              NaN              NaN   
7           NaN          NaN            NaN              NaN              NaN   
8           NaN          NaN            NaN              NaN              NaN   
9           NaN          NaN            NaN              NaN              NaN   

  secondServeIn secondServeZone secondServeXCoord secondServeYCoord isAce  \
0           NaN             NaN               NaN               NaN   NaN   
1           NaN             NaN               NaN               NaN   NaN   
2           NaN             NaN               NaN               NaN   NaN   
3           NaN             NaN               NaN               NaN   NaN   
4           NaN             NaN               NaN               NaN   NaN   
5           NaN             NaN               NaN               NaN   NaN   
6           NaN             NaN               NaN               NaN   NaN   
7           NaN             NaN               NaN               NaN   NaN   
8           NaN             NaN               NaN               NaN   NaN   
9           NaN             NaN               NaN               NaN   NaN   

  shotContactX shotContactY shotDirection shotFhBh isSlice isVolley  \
0          NaN          NaN           NaN      NaN     NaN      NaN   
1          NaN          NaN           NaN      NaN     NaN      NaN   
2          NaN          NaN           NaN      NaN     NaN      NaN   
3          NaN          NaN           NaN      NaN     NaN      NaN   
4          NaN          NaN           NaN      NaN     NaN      NaN   
5          NaN          NaN           NaN      NaN     NaN      NaN   
6          NaN          NaN           NaN      NaN     NaN      NaN   
7          NaN          NaN           NaN      NaN     NaN      NaN   
8          NaN          Na

### isPointStart and isPointEnd columns 

In [39]:
def assign_pointstart(x):
    if (x == 'first_serve') | (x == 'second_serve'):
        return 1
    
    return ''

shot_data['isPointStart'] = swing_data['Type'].apply(assign_pointstart)


index_list = []

for i in swing_data['Point'].unique().tolist():
    last_point_index = swing_data[swing_data['Point'] == i].index[-1]
    index_list.append(last_point_index)
    
shot_data.loc[index_list,'isPointEnd'] = 1
shot_data['isPointEnd'] = shot_data['isPointEnd'].fillna('')

### pointStartTime and pointEndTime Columns

In [40]:
def convert_time(time):
    return int(time * 1000)

# def convert_time(time):
shot_data['pointStartTime'] = swing_data['Video Time'].apply(convert_time)

# Assigns last shot time to pointEndTime column
shot_data['pointEndTime'] = np.where(shot_data['isPointEnd'] == 1, shot_data['pointStartTime'], '')

### pointNumber Column

In [41]:
shot_data['pointNumber'] = swing_data['Point']

### isBreakPoint Column

In [42]:
shot_data['isBreakPoint'] = swing_data['isBreakPoint']

### shotInRally column

In [43]:
shot_data.shotInRally = swing_data.Shot

### side Column

In [44]:
def side(x, side, xcoord):
    if 'deuce' in x:
        return 'Deuce'
    elif 'ad' in x:
        return 'Ad'
    elif 'center_line' in x: # unique values include deuce, ad and center_line
        if (side == 'near') & (xcoord > 0):
            return 'Deuce'
        else:
            return 'Ad'
    else:
        return ''

shot_data['side'] = swing_data.apply(lambda x: side(x['Hit Zone'], x['Hit Side'], x['Bounce (x)']), axis = 1)

### Players

In [45]:
# ucla roster 23-24 men and womens
ucla_roster_23 = ["Gianluca Ballotta", 
                   "Jeffrey Fradkin", 
                   "Alexander Hoogmartens",
                   "Spencer Johnson",
                   "Stefan Leustian",
                   "Timothy Li",
                   "Govind Nanda",
                   "Jorge Plans Gonzalez",
                   "Giacomo Revelli",
                   "Aadarsh Tripathi",
                   "Emon van Loben Sels",
                   "Azuma Visaya",
                   "Rudy Quan",
                   "Leo Von Bismark",
                   
                   "Tian Fangran",
                   "Bianca Fernandez",
                   "Ahmani Guichard",
                   "Kimmi Hance",
                   "Mia Jovic",
                   "Anne-Christine Lutkemeyer",
                   "Vanessa Ong",
                   "Sasha Vagramov",
                   "Elise Wagle"]

In [46]:
# list of names who are playing in match
players = swing_data['Player'].unique()

# checks which one is UCLA player
is_ucla_player = [any([name in roster_name for roster_name in ucla_roster_23]) for name in players]

In [47]:
# assigns ucla player to player 1, and non ucla to player 2
shot_data.loc[0, "player1Name"] = players[is_ucla_player]
shot_data.loc[0, "player2Name"] = players[np.invert(is_ucla_player)]

### serverName Column

In [48]:
def assign_server_name(stroke, server):
    if stroke != 'Serve':
        return ''
    
    if server.startswith(players[is_ucla_player][0]):
        return 'Player1'
    elif server.startswith(players[np.invert(is_ucla_player)][0]):
        return 'Player2'
    
shot_data['serverName'] = swing_data.apply(lambda x: assign_server_name(x['Stroke'], x['Player']), axis=1)
shot_data['serverName'].replace(['', 'na'], pd.NaT, inplace=True)
shot_data['serverName'] = shot_data['serverName'].ffill()

### serverFarNear Column

In [49]:
shot_data.serverFarNear = np.where((swing_data.Stroke == 'Serve'), np.where(swing_data['Hit Side'] == 'far', 'Far', 'Near'), '')
shot_data['serverFarNear'].replace(['', 'na'], pd.NaT, inplace=True)
shot_data['serverFarNear'] = shot_data['serverFarNear'].ffill()

### firstServeIn and secondServeIn Columns

In [50]:
shot_data.firstServeIn = np.where((swing_data.Type == 'first_serve'),np.where((shot_data.isPointStart == 1) & (swing_data['Result'] == 'In'), 1, 0), np.nan)
shot_data.secondServeIn =np.where((swing_data.Type == 'second_serve') & (shot_data.isPointStart == 1), np.where(swing_data['Result'] == 'In', 1,0), np.nan)

### SwingVision Coord Transformation
court coordinates
swing vision - meters, near side center marks (0,0)
singles court x [-4.1148, 4.1148], y [0, 23.7744]
doubles court x [-5.485, 5.485]

our coordinates - center of net (0,0)
singles court x [-157.5, 157.5], y [-455, 455]

shot_x = (157.5/4.1148) * swing_x
shot_y = (455/11.8872) * swing_y + 455
ratio = 38.2764654418

### firstServeXCoord, firstYServeYCoord, secondServeXCoord, and secondServeyCoord Columns

In [51]:
def first_serve_x_coordinates(stroke, x):
    if stroke == 'first_serve':
        return x * 38.2764654418
    else:
        return np.nan

def first_serve_y_coordinates(stroke, y):
    if stroke == 'first_serve':
        return (y - 11.8872) * 38.2764654418
    else:
        return np.nan
    
shot_data['firstServeXCoord'] = swing_data.apply(lambda row: first_serve_x_coordinates(row['Type'], row['Bounce (x)']), axis=1)
shot_data['firstServeYCoord'] = swing_data.apply(lambda row: first_serve_y_coordinates(row['Type'], row['Bounce (y)']), axis=1)


def second_serve_x_coordinates(stroke, x):
    if stroke == 'second_serve':
        return x * 38.2764654418
    else:
        return np.nan

def second_serve_y_coordinates(stroke, y):
    if stroke == 'second_serve':
        return (y - 11.8872) * 38.2764654418
    else:
        return np.nan
    
shot_data['secondServeXCoord'] = swing_data.apply(lambda row: second_serve_x_coordinates(row['Type'], row['Bounce (x)']), axis=1)
shot_data['secondServeYCoord'] = swing_data.apply(lambda row: second_serve_y_coordinates(row['Type'], row['Bounce (y)']), axis=1)

### firstServeZone and secondServeZone Columns
- serving zones: T, Body, Wide
- Wide: x in [-inf, -105] u [105, inf]
- Body: x in [-105, -52.5] u [52.5, 105]
- T: x in [-52.5, 52.5]

In [52]:
def label_zone(x_coord):
    if x_coord != '':
        x_coord = float(x_coord)
        if (x_coord < -105) | (x_coord > 105):
            return 'Wide'
        elif (-105 <= x_coord <= -52.5) | (52.5 <= x_coord <= 105):
            return 'Body'
        elif -52.5 < x_coord < 52.5:
            return 'T'
    return ''

# convert x coord to serve zone
shot_data.firstServeZone = shot_data.firstServeXCoord.apply(label_zone)
shot_data.secondServeZone = shot_data.secondServeXCoord.apply(label_zone)

In [53]:
shot_data.head()

pointScore gameScore setScore isPointStart  pointStartTime isPointEnd  \
0        0-0       0-0      0-0            1            4220              
1        0-0       0-0      0-0                         4990          1   
2       15-0       0-0      0-0            1           26920              
3       15-0       0-0      0-0                        27750              
4       15-0       0-0      0-0                        30870          1   

  pointEndTime  pointNumber isBreakPoint  shotInRally   side serverName  \
0                         1                         1  Deuce    Player1   
1         4990            1                         2  Deuce    Player1   
2                         2                         1     Ad    Player1   
3                         2                         2     Ad    Player1   
4        30870            2                         3     Ad    Player1   

  serverFarNear  firstServeIn firstServeZone  firstServeXCoord  \
0          Near           1.0           Wide       -118.492875   
1          Near           NaN                              NaN   
2          Near           1.0              T         41.603073   
3          Near           NaN                              NaN   
4          Near           NaN                              NaN   

   firstServeYCoord  secondServeIn secondServeZone  secondServeXCoord  \
0        242.945740            NaN                                NaN   
1               NaN            NaN                                NaN   
2        164.843378            NaN                                NaN   
3               NaN            NaN                                NaN   
4               NaN            NaN                                NaN   

   secondServeYCoord isAce shotContactX shotContactY shotDirection shotFhBh  \
0                NaN   NaN          NaN          NaN           NaN      NaN   
1                NaN   NaN          NaN          NaN           NaN      NaN   
2                NaN   NaN          NaN          NaN           NaN      NaN   
3                NaN   NaN          NaN          NaN           NaN      NaN   
4                NaN   NaN          NaN          NaN           NaN      NaN   

  isSlice isVolley isOverhead isApproach isDropshot isLet isExcitingPoint  \
0     NaN      NaN        NaN        NaN        NaN   NaN             NaN   
1     NaN      NaN        NaN        NaN        NaN   NaN             NaN   
2     NaN      NaN        NaN        NaN        NaN   NaN             NaN   
3     NaN      NaN        NaN        NaN        NaN   NaN             NaN   
4     NaN      NaN        NaN        NaN        NaN   NaN             NaN   

  atNetPlayer1 atNetPlayer2 isLob shotLocationX shotLocationY isWinner  \
0          NaN          NaN   NaN           NaN           NaN      NaN   
1          NaN          NaN   NaN           NaN           NaN      NaN   
2          NaN          NaN   NaN           NaN           NaN      NaN   
3          NaN          NaN   NaN           NaN           NaN      NaN   
4          NaN          NaN   NaN           NaN           NaN      NaN   

  isErrorWideR isErrorWideL isErrorNet isErrorLong clientTeam Date Division  \
0          NaN          NaN        NaN         NaN        NaN  NaN      NaN   
1          NaN          NaN        NaN         NaN        NaN  NaN      NaN   
2          NaN          NaN        NaN         NaN        NaN  NaN      NaN   
3          NaN          NaN        NaN         NaN        NaN  NaN      NaN   
4          NaN          NaN        NaN         NaN        NaN  NaN      NaN   

  Event lineupPosition matchDetails matchVenue opponentTeam  \
0   NaN            NaN          NaN        NaN          NaN   
1   NaN            NaN          NaN        NaN          NaN   
2   NaN            NaN          NaN        NaN          NaN   
3   NaN            NaN          NaN        NaN          NaN   
4   NaN            NaN          NaN        NaN          NaN   

         player1Name   play

### isAce Column

In [54]:
shot_data['isAce'] = np.where((swing_data.Stroke == 'Serve') & 
                              (shot_data.isPointEnd == 1) & 
                              (shot_data.secondServeIn != 0), 1, np.nan)

### shotContactX and shotContactY Columns

In [55]:
# Functions to transform the swingvision coordinates
def transf_x_coord_sv_to_shot(sv_col) :
    return sv_col * 38.2764654418
def transf_y_coord_sv_to_shot(sv_col) :
    return (sv_col - 11.8872) * 38.2764654418

# want to convert swingvision coordinates into our own
shot_data['shotContactX'] = transf_x_coord_sv_to_shot(swing_data['Hit (x)'])
shot_data['shotContactY'] = transf_y_coord_sv_to_shot(swing_data['Hit (y)'])

### shotFhBh Column

In [56]:
def classify_shot(stroke):
    if stroke == 'FH Volley':
        return 'Forehand'
    elif stroke == 'BH Volley':
        return 'Backhand'
    elif stroke == 'Forehand':
        return 'Forehand'
    elif stroke == 'Backhand':
        return 'Backhand'
    elif stroke == 'Overhead':
        return 'Forehand'
    else:
        return ''

# Applying the function to the DataFrame
shot_data['shotFhBh'] = swing_data['Stroke'].apply(classify_shot)

### isSlice, isTopspin, isFlat, isKick Columns

In [57]:
shot_data['isSlice'] = swing_data['Spin'].apply(lambda x: '1' if x == 'Slice' else '')
shot_data['isTopspin'] = swing_data['Spin'].apply(lambda x: '1' if x == 'Topspin' else '') # added these metrics
shot_data['isFlat'] = swing_data['Spin'].apply(lambda x: '1' if x == 'Flat' else '') # added these metrics
shot_data['isKick'] = swing_data['Spin'].apply(lambda x: '1' if x == 'Kick' else '') # added these metrics

### isVolley Column

In [58]:
shot_data['isVolley'] = swing_data['Stroke'].apply(lambda x: 1 if x in ['FH Volley', 'BH Volley', 'Volley'] else '') # need to classify shotFhBh when doing isVolley

### isOverhead Column

In [59]:
shot_data['isOverhead'] = swing_data['Stroke'].apply(lambda x: 1 if x == 'Overhead' else '')

### isApproach Column

In [60]:
# maybe run model to predict 

# features to consider:
# player is inside the court

# Workflow:
# watch all points and tag all points that have _____
# subset df with points (testing x and y)


### isDropshot Column

In [61]:
# maybe run model to predict
# features to consider: 
# shotlocationY if close to the net
# shotContactY is close to the net
# speed of the ball (in swingvision data)

### isLet Column

In [62]:
# maybe run model to predict OR get from swingvision data
# features to consider: 


### isExcitingPoint

In [63]:
# maybe run model to predict
# features to consider:
# rally length is long (maybe take _% percintile of rallies)
# point ends in a winner
# either player moves a lot
# amount of volleys, overheads
# breakpoint factor

### atNetPlayer1 and atNetPlayer2 Columns

In [64]:
# aggregated in STP

### isLob Column

In [65]:
# maybe run model to predict
# features to consider:
# opponent is at the net
# speed of the ball (in swingvision data)

### shotLocationX and shotLocationY Columns

In [66]:
# Functions to transform the swingvision coordinates
def transf_x_loc(stroke, sv_col):
    if stroke != 'first_serve' and stroke != 'second_serve':
        return sv_col * 38.2764654418
    return np.nan
    
def transf_y_loc(stroke, sv_col):
    if stroke != 'first_serve' and stroke != 'second_serve':
        return (sv_col - 11.8872) * 38.2764654418
    return np.nan

# want to convert swingvision coordinates into our own
shot_data['shotLocationX'] = swing_data.apply(lambda x: transf_x_loc(x['Type'], x['Bounce (x)']), axis=1)
shot_data['shotLocationY'] = swing_data.apply(lambda x: transf_y_loc(x['Type'], x['Bounce (y)']), axis=1)

### shotDirection column

In [67]:
# down the line --> switches btwn deuce and ad
# crosscourt --> remains on same side
shot_data['shotDirection'] = np.where((shot_data.shotContactX * shot_data.shotLocationX > 0) & (shot_data.shotInRally != 1), 
"Down the Line", 
    np.where((shot_data.shotInRally != 1), 'Crosscourt', ''))

### isWinner Column

In [68]:
shot_data.isWinner = np.where((shot_data.isPointEnd == 1) & (shot_data.secondServeIn != '0') &
                              (swing_data.Result == 'In'), 1, np.nan)


### isErrorWideR Column

In [69]:
def wide_right_function(side, x, y, end):
    if (side == 'far' and x < -157.5 and end == 1) or (side == 'near' and x > 157.5 and end == 1):
        return 1
    return np.nan

# Assign 'isErrorWideR' using values from both 'swing_data' and 'shot_data'
shot_data['isErrorWideR'] = shot_data.apply(lambda x: wide_right_function(swing_data.loc[x.name, 'Hit Side'], 
                                                                x['shotLocationX'], x['shotLocationY'], x['isPointEnd']), axis=1)


### isErrorWideL Column

In [70]:
def wide_left_function(side, x, y, end):
    if (side == 'far' and x > 157.5 and end == 1) or (side == 'near' and x < -157.5 and end == 1):
        return 1
    return np.nan

# Assign 'isErrorWideR' using values from both 'swing_data' and 'shot_data'
shot_data['isErrorWideL'] = shot_data.apply(lambda x: wide_left_function(swing_data.loc[x.name, 'Hit Side'], 
                                                                x['shotLocationX'], x['shotLocationY'], x['isPointEnd']), axis=1)


### isErrorNet Column

In [71]:
shot_data.isErrorNet = np.where((swing_data.Result == 'Net'), 1, np.nan)

### isErrorLong Column

In [72]:
shot_data['isErrorLong'] = np.where((swing_data['Result'] == 'Out') & (shot_data['shotLocationY'].abs() > 455), 1, np.nan)

### Group First Serve and Second Serve Columns

In [73]:
# All columns
default_cols = ['pointScore', 'gameScore', 'setScore', 'isPointStart', 'pointStartTime',
       'isPointEnd', 'pointEndTime', 'pointNumber', 'isBreakPoint',
       'shotInRally', 'side', 'serverName', 'serverFarNear', 'firstServeIn',
       'firstServeZone', 'firstServeXCoord', 'firstServeYCoord',
       'secondServeIn', 'secondServeZone', 'secondServeXCoord',
       'secondServeYCoord', 'isAce', 'shotContactX', 'shotContactY',
       'shotDirection', 'shotFhBh', 'isSlice', 'isVolley', 'isOverhead',
       'isApproach', 'isDropshot', 'isExcitingPoint', 'atNetPlayer1',
       'atNetPlayer2', 'isLob', 'shotLocationX', 'shotLocationY', 'isWinner',
       'isErrorWideR', 'isErrorWideL', 'isErrorNet', 'isErrorLong',
       'clientTeam', 'Date', 'Division', 'Event', 'lineupPosition',
       'matchDetails', 'matchVenue', 'opponentTeam', 'player1Name',
       'player2Name', 'player1Hand', 'player2Hand', 'Round', 'Surface',
       'Notes', 'isTopspin', 'isFlat', 'isKick']

# Assign all columns to have value be taken from the first serve row
agg_dict = {col: 'first' for col in default_cols}

# Reassign select columns to have value be taken form the second serve row
agg_dict.update({'isPointEnd': 'last', 
                 'pointEndTime': 'last', 
                 'secondServeIn' : 'last',
                 'secondServeZone' : 'last',
                 'secondServeXCoord' : 'last', 
                 'secondServeYCoord' : 'last',
                 'isAce' : 'last', 
                 'shotContactX' : 'last', 
                 'shotContactY' : 'last',
                 'isWinner' : 'last',
                 'isErrorWideR' : 'last',
                 'isErrorWideL' : 'last', 
                 'isErrorNet' : 'last',
                 'isErrorLong' : 'last'
                })

# Group by isPointStart and pointNumber
grouped_df = shot_data.groupby(['shotInRally', 'pointNumber'], as_index=False).agg(agg_dict)
shot_data = grouped_df.sort_values(by=['pointNumber', 'shotInRally'], ascending=[True, True]).reset_index(drop = True)

### Save as CSV

In [74]:
player1NameNoSpace = str(shot_data.iloc[0]['player1Name']).replace(" ", "")
player2NameNoSpace = str(shot_data.iloc[0]['player2Name']).replace(" ", "")

shot_data.to_csv(f'swingvision_{player1NameNoSpace}_{player2NameNoSpace}.csv', index=False)
print(f'swingvision_{player1NameNoSpace}_{player2NameNoSpace}.csv')

swingvision_GianlucaBallotta_MaxWestphal.csv


### Notebook End

## Errors in Swingvision Data Exploration

#### Chcek all the rows where isPointEnd != 1 and there is  isWinner, isErrorWideL, isErrorWideR, isErrorNet, isErrorLong

In [75]:
point_error = shot_data[(shot_data['isPointEnd'] != 1) & (shot_data['isPointStart'] != 1) &
          ((shot_data['isWinner'] == 1) | 
          (shot_data['isErrorNet'] == 1) | 
          (shot_data['isErrorLong'] == 1) |
          (shot_data['isErrorWideL'] == 1) |
          (shot_data['isErrorWideR'] == 1))]

point_error_numbers = point_error['pointNumber'].to_list()

if len(point_error) > 0:
    display(point_error)
    raise ValueError('Manually check points', point_error_numbers)

pointScore gameScore setScore isPointStart  pointStartTime isPointEnd  \
71        0-40       1-2      0-0                       820179              
100      30-15       2-2      0-0                       985369              
117       0-15       3-2      0-0                      1215510              
175      40-15       3-3      0-0                      1620050              
290       0-15       0-2      0-1                      2899429              
392      40-30       1-4      0-1                      3819719              
425      40-15       1-5      0-1                      4046399              

    pointEndTime  pointNumber isBreakPoint  shotInRally   side serverName  \
71                         22            1            2     Ad    Player2   
100                        26                         3  Deuce    Player1   
117                        31                         2     Ad    Player2   
175                        41                         4  Deuce    Player1   
290                        68                         2     Ad    Player1   
392                        90                         5     Ad    Player2   
425                        96                         9  Deuce    Player1   

    serverFarNear  firstServeIn firstServeZone  firstServeXCoord  \
71            Far           NaN                              NaN   
100          Near           NaN                              NaN   
117          Near           NaN                              NaN   
175           Far           NaN                              NaN   
290          Near           NaN                              NaN   
392           Far           NaN                              NaN   
425          Near           NaN                              NaN   

     firstServeYCoord  secondServeIn secondServeZone  secondServeXCoord  \
71                NaN            NaN                                NaN   
100               NaN            NaN                                NaN   
117               NaN            NaN                                NaN   
175               NaN            NaN                                NaN   
290               NaN            NaN                                NaN   
392               NaN            NaN                                NaN   
425               NaN            NaN                                NaN   

     secondServeYCoord  isAce  shotContactX  shotContactY  shotDirection  \
71                 NaN    NaN   -179.484853   -503.735165  Down the Line   
100                NaN    NaN      8.295237   -517.557715     Crosscourt   
117                NaN    NaN    222.492252    492.152056  Down the Line   
175                NaN    NaN    140.848589   -539.831173  Down the Line   
290                NaN    NaN    196.366803    499.997315     Crosscourt   
392                NaN    NaN     57.875776    305.829648     Crosscourt   
425                NaN    NaN     47.024092   -438.431802  Down the Line   

     shotFhBh isSlice isVolley isOverhead isApproach isDropshot  \
71   Backhand                                   None       None   
100  Backhand                                   None       None   
117  Backhand                                   None       None   
175  Forehand                                   None       None   
290  Forehand                                   None       None   
392                          1                  None       None   
425  Backhand                                   None       None   

    isExcitingPoint atNetPlayer1 atNetPlayer2 isLob  shotLocationX  \
71             None         None         None  None     -47.275454   
100            None         None         None  None     -98.599142   
117            None         None         None  None     187.943914   
175            None         None         None  None      65.017476   
290            None         None         None  None    -142.504506   
392            None         None         None  None      -

ValueError: ('Manually check points', [22, 26, 31, 41, 68, 90, 96])

In [76]:
shot_data[shot_data['pointNumber'] == 2]

pointScore gameScore setScore isPointStart  pointStartTime isPointEnd  \
3       15-0       0-0      0-0            1           35349          1   

  pointEndTime  pointNumber isBreakPoint  shotInRally side serverName  \
3        69970            2                         1   Ad    Player1   

  serverFarNear  firstServeIn firstServeZone  firstServeXCoord  \
3          Near           0.0           Body         99.205173   

   firstServeYCoord  secondServeIn secondServeZone  secondServeXCoord  \
3        -42.625667            0.0               T            3.78294   

   secondServeYCoord  isAce  shotContactX  shotContactY shotDirection  \
3        -182.774524    NaN    -13.142607   -442.920292                 

  shotFhBh isSlice isVolley isOverhead isApproach isDropshot isExcitingPoint  \
3                                            None       None            None   

  atNetPlayer1 atNetPlayer2 isLob  shotLocationX  shotLocationY  isWinner  \
3         None         None  None            NaN            NaN       NaN   

   isErrorWideR  isErrorWideL  isErrorNet  isErrorLong clientTeam  Date  \
3           NaN           NaN         1.0          NaN       None  None   

  Division Event lineupPosition matchDetails matchVenue opponentTeam  \
3     None  None           None         None       None         None   

  player1Name player2Name player1Hand player2Hand Round Surface Notes  \
3        None        None        None        None  None    None  None   

  isTopspin isFlat isKick  
3                1

#### Check all the rows where there is isPointEnd == 1 but there is no isWinner, isErrorWideL, isErrorWideR, isErrorNet, isErrorLong
- Cj reccomendation: have this error check autmatically fill in how the point ends based on coordinate data

In [69]:
point_error = shot_data[(shot_data['isPointEnd'] == 1) &
                          (shot_data['isWinner'] != 1) &
                          (shot_data['isErrorWideL'] != 1) &
                          (shot_data['isErrorWideR'] != 1) &
                          (shot_data['isErrorNet'] != 1) & 
                          (shot_data['isErrorLong'] != 1) &
                          (shot_data['firstServeIn'] != 0) & 
                          (shot_data['secondServeIn'] != 0)]

point_error_numbers = point_error['pointNumber'].to_list()

if point_error.empty:
    print('Check Passed ✓')
else:

    display(point_error)
    raise ValueError('Manually check points', point_error_numbers)

Check Passed ✓


#### Volleys


In [91]:
shot_data.query('isVolley == 1')

pointScore gameScore setScore isPointStart  pointStartTime isPointEnd  \
13       15-40       0-0      0-0                       166270              
99        15-0       1-3      0-0                       870559          1   
105      15-15       1-3      0-0                       926859          1   
164        0-0       2-5      0-0                      1631349              
166        0-0       2-5      0-0                      1634449          1   
196       0-15       0-0      0-1                      2076939          1   
212        0-0       0-1      0-1                      2317840          1   
217       15-0       0-1      0-1                      2355540              
219       15-0       0-1      0-1                      2356939              
307      30-30       1-3      0-1                      3171030              
362       15-0       2-4      0-1                      3676419              
385      30-30       2-4      0-1                      3898219          1   

    pointEndTime  pointNumber isBreakPoint  shotInRally   side serverName  \
13                          5            1            5  Deuce    Player1   
99        870559           21                        13  Deuce    Player1   
105       926859           22                         6     Ad    Player1   
164                        35                         4     Ad    Player2   
166      1634449           35                         6     Ad    Player2   
196      2076939           41                         5     Ad    Player1   
212      2317840           46                         4  Deuce    Player2   
217                        47                         5  Deuce    Player2   
219                        47                         7  Deuce    Player2   
307                        67                         3     Ad    Player1   
362                        78                         6  Deuce    Player1   
385      3898219           81                         6     Ad    Player2   

    serverFarNear  firstServeIn firstServeZone  firstServeXCoord  \
13           Near           NaN                              NaN   
99           Near           NaN                              NaN   
105          Near           NaN                              NaN   
164           Far           NaN                              NaN   
166           Far           NaN                              NaN   
196          Near           NaN                              NaN   
212          Near           NaN                              NaN   
217          Near           NaN                              NaN   
219          Near           NaN                              NaN   
307          Near           NaN                              NaN   
362           Far           NaN                              NaN   
385           Far           NaN                              NaN   

     firstServeYCoord  secondServeIn secondServeZone  secondServeXCoord  \
13                NaN            NaN                                NaN   
99                NaN            NaN                                NaN   
105               NaN            NaN                                NaN   
164               NaN            NaN                                NaN   
166               NaN            NaN                                NaN   
196               NaN            NaN                                NaN   
212               NaN            NaN                                NaN   
217               NaN            NaN                                NaN   
219               NaN            NaN                                NaN   
307               NaN            NaN                                NaN   
362               NaN            NaN                                NaN   
385               NaN            NaN                                NaN   

     secondServeYCoord  isAce  shotContactX  shotContactY  shotDirection  \
13                 NaN    NaN     30.445254    -41.593313  Down

In [93]:
shot_data[355:500]

pointScore gameScore setScore isPointStart  pointStartTime isPointEnd  \
355        0-0       2-4      0-1            1         3623590              
356        0-0       2-4      0-1                      3624419          1   
357       15-0       2-4      0-1            1         3650889              
358       15-0       2-4      0-1                      3670639              
359       15-0       2-4      0-1                      3672020              
360       15-0       2-4      0-1                      3673169              
361       15-0       2-4      0-1                      3674750              
362       15-0       2-4      0-1                      3676419              
363       15-0       2-4      0-1                      3677389          1   
364      15-15       2-4      0-1            1         3717419              
365      15-15       2-4      0-1                      3718120              
366      15-15       2-4      0-1                      3719419              
367      15-15       2-4      0-1                      3721370              
368      15-15       2-4      0-1                      3722620              
369      15-15       2-4      0-1                      3724070              
370      15-15       2-4      0-1                      3725219              
371      15-15       2-4      0-1                      3726540              
372      15-15       2-4      0-1                      3727949              
373      15-15       2-4      0-1                      3729300              
374      15-15       2-4      0-1                      3730419              
375      15-15       2-4      0-1                      3731719              
376      15-15       2-4      0-1                      3733120          1   
377      15-30       2-4      0-1            1         3771649              
378      15-30       2-4      0-1                      3772540              
379      15-30       2-4      0-1                      3774550          1   
380      30-30       2-4      0-1            1         3860719              
381      30-30       2-4      0-1                      3887540              
382      30-30       2-4      0-1                      3889219              
383      30-30       2-4      0-1                      3890300              
384      30-30       2-4      0-1                      3897139              
385      30-30       2-4      0-1                      3898219          1   
386      30-40       2-4      0-1            1         3912050              
387      30-40       2-4      0-1                      3912800              
388      30-40       2-4      0-1                      3914419              
389      30-40       2-4      0-1                      3915370              
390      30-40       2-4      0-1                      3917020          1   
391        0-0       2-5      0-1            1         3939489              
392        0-0       2-5      0-1                      3955530              
393        0-0       2-5      0-1                      3956719              
394        0-0       2-5      0-1                      3958030          1   
395       15-0       2-5      0-1            1         3978419              
396       15-0       2-5      0-1                      3979199              
397       15-0       2-5      0-1                      3981120          1   
398       30-0       2-5      0-1            1         4004419              
399       30-0       2-5      0-1                      4015229              
400       30-0       2-5      0-1                      4016649          1   
401       40-0       2-5      0-1            1         4045479              
402       40-0       2-5      0-1                      4046100              
403       40-0       2-5      0-1                      4047879          1   

    pointEndTime  pointNumber isBreakPoint  shotInRally   side serverName  \
355                        77                         1  Deuce    Player1   

#### Overheads

In [72]:
shot_data.query('isOverhead == 1')

pointScore gameScore setScore isPointStart  pointStartTime isPointEnd  \
221       15-0       0-1      0-1                      2360239          1   

    pointEndTime  pointNumber isBreakPoint  shotInRally   side serverName  \
221      2360239           47                         9  Deuce    Player2   

    serverFarNear  firstServeIn firstServeZone  firstServeXCoord  \
221          Near           NaN                              NaN   

     firstServeYCoord  secondServeIn secondServeZone  secondServeXCoord  \
221               NaN            NaN                                NaN   

     secondServeYCoord  isAce  shotContactX  shotContactY shotDirection  \
221                NaN    NaN    118.944193   -252.597649    Crosscourt   

     shotFhBh isSlice isVolley isOverhead isApproach isDropshot  \
221  Forehand                           1       None       None   

    isExcitingPoint atNetPlayer1 atNetPlayer2 isLob  shotLocationX  \
221            None         None         None  None    -160.746265   

     shotLocationY  isWinner  isErrorWideR  isErrorWideL  isErrorNet  \
221      454.59982       NaN           NaN           1.0         NaN   

     isErrorLong clientTeam  Date Division Event lineupPosition matchDetails  \
221          NaN       None  None     None  None           None         None   

    matchVenue opponentTeam player1Name player2Name player1Hand player2Hand  \
221       None         None        None        None        None        None   

    Round Surface Notes isTopspin isFlat isKick  
221  None    None  None                1

#### Aces
- WARNING: Not accurate
- FIX: counts double faults as aces

In [73]:
shot_data.query('isAce == 1')

pointScore gameScore setScore isPointStart  pointStartTime isPointEnd  \
79       30-0       1-2      0-0            1          775929          1   

   pointEndTime  pointNumber isBreakPoint  shotInRally   side serverName  \
79       775929           18                         1  Deuce    Player2   

   serverFarNear  firstServeIn firstServeZone  firstServeXCoord  \
79           Far           1.0              T          1.201996   

    firstServeYCoord  secondServeIn secondServeZone  secondServeXCoord  \
79        -227.85134            NaN                                NaN   

    secondServeYCoord  isAce  shotContactX  shotContactY shotDirection  \
79                NaN    1.0    -48.704735    429.271479                 

   shotFhBh isSlice isVolley isOverhead isApproach isDropshot isExcitingPoint  \
79                                            None       None            None   

   atNetPlayer1 atNetPlayer2 isLob  shotLocationX  shotLocationY  isWinner  \
79         None         None  None            NaN            NaN       1.0   

    isErrorWideR  isErrorWideL  isErrorNet  isErrorLong clientTeam  Date  \
79           NaN           NaN         NaN          NaN       None  None   

   Division Event lineupPosition matchDetails matchVenue opponentTeam  \
79     None  None           None         None       None         None   

   player1Name player2Name player1Hand player2Hand Round Surface Notes  \
79        None        None        None        None  None    None  None   

   isTopspin isFlat isKick  
79                1

#### Double Faults

In [88]:
shot_data.query('firstServeIn == 0').query('secondServeIn == 0')

pointScore gameScore setScore isPointStart  pointStartTime isPointEnd  \
3         15-0       0-0      0-0            1           35349          1   
34       40-15       0-1      0-0            1          370380              
106      30-15       1-3      0-0            1          955270          1   
265      40-40       0-2      0-1            1         2710100          1   

    pointEndTime  pointNumber isBreakPoint  shotInRally   side serverName  \
3          69970            2                         1     Ad    Player1   
34                         10                         1  Deuce    Player2   
106       964309           23                         1     Ad    Player1   
265      2719899           57            1            1     Ad    Player1   

    serverFarNear  firstServeIn firstServeZone  firstServeXCoord  \
3            Near           0.0           Body         99.205173   
34           Near           0.0              T         11.307059   
106          Near           0.0           Body         97.435154   
265           Far           0.0           Wide       -191.688807   

     firstServeYCoord  secondServeIn secondServeZone  secondServeXCoord  \
3          -42.625667            0.0               T           3.782940   
34         -25.638840            0.0            Body         -52.542295   
106        -42.945888            0.0               T          33.428560   
265       -220.564955            0.0               T          47.150252   

     secondServeYCoord  isAce  shotContactX  shotContactY shotDirection  \
3          -182.774524    NaN    -13.142607   -442.920292                 
34          263.234181    NaN     49.729741   -446.557169                 
106         -44.775312    NaN    -28.205430   -471.140573                 
265        -248.594198    NaN     32.941262    456.425760                 

    shotFhBh isSlice isVolley isOverhead isApproach isDropshot  \
3                                              None       None   
34                                             None       None   
106                1                           None       None   
265                                            None       None   

    isExcitingPoint atNetPlayer1 atNetPlayer2 isLob  shotLocationX  \
3              None         None         None  None            NaN   
34             None         None         None  None            NaN   
106            None         None         None  None            NaN   
265            None         None         None  None            NaN   

     shotLocationY  isWinner  isErrorWideR  isErrorWideL  isErrorNet  \
3              NaN       NaN           NaN           NaN         1.0   
34             NaN       NaN           NaN           NaN         1.0   
106            NaN       NaN           NaN           NaN         1.0   
265            NaN       NaN           NaN           NaN         NaN   

     isErrorLong clientTeam  Date Division Event lineupPosition matchDetails  \
3            NaN       None  None     None  None           None         None   
34           NaN       None  None     None  None           None         None   
106          NaN       None  None     None  None           None         None   
265          NaN       None  None     None  None           None         None   

    matchVenue opponentTeam player1Name player2Name player1Hand player2Hand  \
3         None         None        None        None        None        None   
34        None         None        None        None        None        None   
106       None         None        None        None        None        None   
265       None         None        None        None        None        None   

    Round Surface Notes isTopspin isFlat isKick  
3    None    None  None                1         
34   None    None  None                1         
106  None    None  None                          
265  None    None  None                1

In [85]:
shot_data[shot_data['pointNumber'] == 10]

pointScore gameScore setScore isPointStart  pointStartTime isPointEnd  \
34      40-15       0-1      0-0            1          370380              
35      40-15       0-1      0-0                       383809              
36      40-15       0-1      0-0                       384850              
37      40-15       0-1      0-0                       386179              
38      40-15       0-1      0-0                       387649              
39      40-15       0-1      0-0                       388850              
40      40-15       0-1      0-0                       390079              
41      40-15       0-1      0-0                       391660          1   

   pointEndTime  pointNumber isBreakPoint  shotInRally   side serverName  \
34                        10                         1  Deuce    Player2   
35                        10                         2  Deuce    Player2   
36                        10                         3  Deuce    Player2   
37                        10                         4  Deuce    Player2   
38                        10                         5  Deuce    Player2   
39                        10                         6  Deuce    Player2   
40                        10                         7  Deuce    Player2   
41       391660           10                         8     Ad    Player2   

   serverFarNear  firstServeIn firstServeZone  firstServeXCoord  \
34          Near           0.0              T         11.307059   
35          Near           NaN                              NaN   
36          Near           NaN                              NaN   
37          Near           NaN                              NaN   
38          Near           NaN                              NaN   
39          Near           NaN                              NaN   
40          Near           NaN                              NaN   
41          Near           NaN                              NaN   

    firstServeYCoord  secondServeIn secondServeZone  secondServeXCoord  \
34         -25.63884            0.0            Body         -52.542295   
35               NaN            NaN                                NaN   
36               NaN            NaN                                NaN   
37               NaN            NaN                                NaN   
38               NaN            NaN                                NaN   
39               NaN            NaN                                NaN   
40               NaN            NaN                                NaN   
41               NaN            NaN                                NaN   

    secondServeYCoord  isAce  shotContactX  shotContactY  shotDirection  \
34         263.234181    NaN     49.729741   -446.557169                  
35                NaN    NaN    -73.238955    498.070669     Crosscourt   
36                NaN    NaN     15.186647   -469.531966     Crosscourt   
37                NaN    NaN    -73.286341    512.967984     Crosscourt   
38                NaN    NaN    115.323239   -518.601132     Crosscourt   
39                NaN    NaN   -176.819434    536.631299     Crosscourt   
40                NaN    NaN     99.680184   -492.940322  Down the Line   
41                NaN    NaN     44.211806    479.083858     Crosscourt   

    shotFhBh isSlice isVolley isOverhead isApproach isDropshot  \
34                                             None       None   
35  Backhand                                   None       None   
36  Forehand                                   None       None   
37  Forehand                                   None       None   
38  Forehand                                   None       None   
39  Forehand                                   None       None   
40  Forehand                                   None       None   
41  Backhand                                   None       None   

   isExcitingPoint atNetPlayer1 atNetPlayer2 isLob  shotLocationX  \
34            None    

### Check all points where double fault occurs (firstServeIn == 0 & secondServeIn == 0) but len(shotInRally) > 1
- Check double fault but the point continues

### Check all the points where everytime the server changes, the first pointScore should be "0-0". If not output error
- Govind Nanda vs Cooper Williams (Harvard) row 380

### Points

In [ ]:
# # ad scoring?

# ad_scoring = False

In [ ]:
# # want to record the score every time a point ends
# # points: server - returner
# # games: ucla (player1) - opp
# # sets: ucla (player1) - opp
# points = np.zeros(2)
# games = np.zeros(2)
# sets = np.zeros(2)
# pt_values = [0, 15, 30, 40]



# shot_data.loc[0,"pointScore"] = f"{pt_values[int(points[0])]} - {pt_values[int(points[1])]}"
# shot_data.loc[0,"gameScore"] = f"{games[0]} - {games[1]}"
# shot_data.loc[0,"setScore"] = f"{sets[0]} - {sets[1]}"

# shot_data["isBreakPoint"] = ''

# error_cols = [x for x in shot_data.columns if "isError" in x]

# for i in range(0, len(shot_data.pointScore) - 1):
#     if shot_data.loc[i+1, "isPointStart"] == 1: # means we gotta update pts
#         # determine point score by checking last shot
#         if shot_data.loc[i, "isWinner"] == "1":
#             # check if player 1 or 2 won pt
#             pt_winner_player_num = (np.where(shot_data.loc[i, "Player"] == shot_data.loc[0,"player1Name"], '1', '2'))
#         elif shot_data.loc[i, 'secondServeIn'] == "0": # double fault
#             pt_winner_player_num = (np.where(shot_data.loc[i, "Player"] == shot_data.loc[0,"player1Name"], '2', '1'))
#         elif any(shot_data.loc[i,error_cols] == "1"):
#             # winner is the player who did NOT hit that shot
#             pt_winner_player_num = (np.where(shot_data.loc[i, "Player"] == shot_data.loc[0,"player1Name"], '2', '1'))
#         else:
#             print("no pt recorded at row ", i)

#         if shot_data.loc[i, "serverName"] is not None:
#             didServerWinPt = shot_data.loc[i, "serverName"][-1] == pt_winner_player_num
#         else:
#             print(f"Server name is None at row {i}. Skipping this point.")
#             continue  # Skip this point if server name is None
        
#         if didServerWinPt:
#             points[0] += 1
#         else:
#             points[1] += 1


#         if ad_scoring: # checks if need to win by 2
#             if any(points > 3) and abs(points[0] - points[1]) >= 2:
#                 game_winner = np.argmax(points)  # Find who won the game
#                 games[game_winner] += 1
#                 points = np.zeros(2)  # Reset point values
#             if any(games > 5) and abs(games[0] - games[1]) >= 2:
#                 set_winner = np.argmax(games)  # Find who won the set
#                 sets[set_winner] += 1
#                 games = np.zeros(2)  # Reset game values   
#         else:
#             if points[1] == 3: # if the returner has 40 pts and can win the game
#                 shot_data.loc[i + 1, 'isBreakPoint'] = '1'
#             if any(points > 3):
#                 game_winner = np.argmax(points)  # Find who won the game
#                 games[game_winner] += 1
#                 points = np.zeros(2)  # Reset point values
#             if any(games > 5):
#                 set_winner = np.argmax(games)  # Find who won the set
#                 sets[set_winner] += 1
#                 games = np.zeros(2)  # Reset game values   

#     # Update the scores in the shot_data DataFrame
#     shot_data.loc[i+1,"pointScore"] = f"{pt_values[int(points[0])]} - {pt_values[int(points[1])]}"
#     shot_data.loc[i+1,"gameScore"] = f"{int(games[0])} - {int(games[1])}"
#     shot_data.loc[i+1,"setScore"] = f"{int(sets[0])} - {int(sets[1])}"


# # Additional comments for further updates:
# # - Tiebreak scenarios are not yet handled and need to be accounted for in future versions.